pinot-core/src/main/java/org/apache/pinot/core/util/TableConfigUtils.java

```diff
@@ -230,6 +234,38 @@ private static void validateIngestionConfig(@Nullable IngestionConfig ingestionC
     }
   }
 
+  /**
+   * Validates the upsert-related configurations
+   *
+   */
+  private static void validateUpsertConfig(TableConfig tableConfig, Schema schema) {
+    if (tableConfig.getUpsertMode() == UpsertConfig.Mode.NONE) {
+      return;
+    }
+    // primary key exists
+    if (schema.getPrimaryKeyColumns() == null || schema.getPrimaryKeyColumns().isEmpty()) {
+      throw new IllegalStateException("Upsert table must have primary key columns in the schema.");
+    }
+    // replica group is configured for routing
+    if (tableConfig.getRoutingConfig() == null || !tableConfig.getRoutingConfig().getInstanceSelectorType()
+        .equalsIgnoreCase(RoutingConfig.REPLICA_GROUP_INSTANCE_SELECTOR_TYPE)) {
+      throw new IllegalStateException("Upsert table must use replicaGroup as the routing config.");
+    }
```

Comment:
This can potentially have NPE
```suggestion
    Preconditions.checkState(tableConfig.getRoutingConfig() != null && RoutingConfig.REPLICA_GROUP_INSTANCE_SELECTOR_TYPE. equalsIgnoreCase(tableConfig.getRoutingConfig().getInstanceSelectorType()), "Upsert table must use replica-group based routing")
```

Reply:
why? this seems the same rewrite.

Reply:
`tableConfig.getRoutingConfig().getInstanceSelectorType()` might be `null`

Reply:
then it's the same :)